include/zephyr/logging/log.h

```diff
@@ -89,13 +90,346 @@ extern "C" {
  */
 #define LOG_WRN_ONCE(...)					\
 	do {							\
-		static uint8_t __warned;			\
-		if (unlikely(__warned == 0)) {			\
+		static atomic_t __warned;			\
+		if (unlikely(atomic_cas(&__warned, 0, 1))) {	\
 			Z_LOG(LOG_LEVEL_WRN, __VA_ARGS__);	\
-			__warned = 1;				\
 		}						\
 	} while (0)
 
+#if defined(CONFIG_LOG_RATELIMIT) || defined(__DOXYGEN__)
+/**
+ * @brief Core rate-limited logging macro
+ *
+ * @details Internal macro that provides rate-limited logging functionality.
+ * Uses atomic operations to ensure thread safety in multi-threaded environments.
+ * Only one thread can successfully log a message within the specified rate limit period.
+ *
+ * @param _level Log level (LOG_LEVEL_ERR, LOG_LEVEL_WRN, etc.)
+ * @param _rate_ms Minimum interval in milliseconds between log messages.
+ * @param _skipped_msg Message to show when messages are skipped.
+ * @param _log_func Logging function to call (Z_LOG or Z_LOG_HEXDUMP)
+ * @param ... Arguments to pass to the logging function.
+ */
+#define _LOG_RATELIMIT_CORE(_level, _rate_ms, _skipped_msg, _log_func, ...)	\
+	do {									\
+		static atomic_t __last_log_time ;		\
```

Comment:
can you add support for CONFIG_LOG_RATELIMIT_FALLBACK_DROP and  CONFIG_LOG_RATELIMIT inside this macro as `if(IS_ENABLED( CONFIG_LOG_RATELIMIT) ...` instead of using ifdefs. That is the approach in all logging macros. It ensures that there is no warning about unused variables when logging is disabled and variables are only used in logging macros.

Reply:
No, `COND_CODE_1(a, (b), (c))` is resolved by preprocessor and only `b` or `c` is seen by the compiler. When `if (IS_ENABLED(a) b else c` is used then both `b` and `c` are compiled (but one is cut).